gradle/instrumentation.gradle

```diff
@@ -55,3 +51,73 @@ afterEvaluate {
     project(':javaagent-tooling').configurations.instrumentationMuzzle + configurations.runtimeClasspath
   ).configure()
 }
+
+configurations {
+  testInstrumentation
+}
+
+shadowJar {
+  configurations = [project.configurations.runtimeClasspath, project.configurations.testInstrumentation]
+  mergeServiceFiles()
+
+  archiveFileName = 'agent-testing.jar'
+
+  dependencies {
+    exclude(project(':javaagent-bootstrap'))
+    exclude(project(':instrumentation-api'))
+    exclude(project(':javaagent-api'))
+    exclude(project(':javaagent-tooling'))
+    exclude(dependency(deps.bytebuddy))
+    exclude(dependency(deps.opentelemetryApi))
+    exclude(dependency(deps.opentelemetryContext))
+    exclude(dependency(deps.slf4j))
+  }
+
+  // rewrite library instrumentation dependencies
+  relocate "io.opentelemetry.instrumentation", "io.opentelemetry.javaagent.shaded.instrumentation"
+
+  // Prevents conflict with other SLF4J instances. Important for premain.
+  relocate 'org.slf4j', 'io.opentelemetry.javaagent.slf4j'
+  // rewrite dependencies calling Logger.getLogger
+  relocate 'java.util.logging.Logger', 'io.opentelemetry.javaagent.bootstrap.PatchLogger'
+
+  // prevents conflict with library instrumentation
+  relocate 'io.opentelemetry.instrumentation.api', 'io.opentelemetry.javaagent.shaded.instrumentation.api'
+
+  // relocate OpenTelemetry API usage
+  relocate "io.opentelemetry.api", "io.opentelemetry.javaagent.shaded.io.opentelemetry.api"
+  relocate "io.opentelemetry.spi", "io.opentelemetry.javaagent.shaded.io.opentelemetry.spi"
+  relocate "io.opentelemetry.context", "io.opentelemetry.javaagent.shaded.io.opentelemetry.context"
+
+  // this is for instrumentation on opentelemetry-api itself
+  relocate "application.io.opentelemetry", "io.opentelemetry"
+}
+
+tasks.withType(Test).configureEach {
+  // Multi-threaded logging seems to be causing deadlocks with Gradle's log capture.
+  //  jvmArgs "-Dio.opentelemetry.javaagent.slf4j.simpleLogger.defaultLogLevel=debug"
+  //  jvmArgs "-Dorg.slf4j.simpleLogger.defaultLogLevel=debug"
```

Comment:
this is a bummer, see similar comment in `javaagent.gradle`

here's a deadlock I captured locally (interestingly doesn't deadlock on windows, but does on linux):

```
Found one Java-level deadlock:
=============================
"main":
  waiting to lock monitor 0x00007f1b64008580 (object 0x00000000fd873d20, a java.lang.Object),
  which is held by "IntervalMetricReader-1"
"IntervalMetricReader-1":
  waiting for ownable synchronizer 0x00000000fba0b0b0, (a java.util.concurrent.locks.ReentrantLock$NonfairSync),
  which is held by "main"

Java stack information for the threads listed above:
===================================================
"main":
        at org.gradle.internal.logging.sink.OutputEventRenderer.onOutput(OutputEventRenderer.java:431)
        - waiting to lock <0x00000000fd873d20> (a java.lang.Object)
        at org.gradle.internal.logging.sink.OutputEventListenerManager$1.onOutput(OutputEventListenerManager.java:36)
        at org.gradle.internal.logging.services.TextStreamOutputEventListener.onTextEvent(TextStreamOutputEventListener.java:58)
        at org.gradle.internal.logging.services.TextStreamOutputEventListener.onOutput(TextStreamOutputEventListener.java:42)
        at org.gradle.internal.logging.source.PrintStreamLoggingSystem$OutputEventDestination.onOutput(PrintStreamLoggingSystem.java:173)
        at org.gradle.internal.logging.source.PrintStreamLoggingSystem$1.text(PrintStreamLoggingSystem.java:45)
        at org.gradle.internal.io.LineBufferingOutputStream.flush(LineBufferingOutputStream.java:98)
        at org.gradle.internal.io.LineBufferingOutputStream.write(LineBufferingOutputStream.java:82)
        at java.io.OutputStream.write(java.base@11.0.4/OutputStream.java:157)
        at java.io.PrintStream.write(java.base@11.0.4/PrintStream.java:559)
        - locked <0x00000000faf18930> (a java.io.PrintStream)
        at sun.nio.cs.StreamEncoder.writeBytes(java.base@11.0.4/StreamEncoder.java:233)
        at sun.nio.cs.StreamEncoder.implFlushBuffer(java.base@11.0.4/StreamEncoder.java:312)
        at sun.nio.cs.StreamEncoder.flushBuffer(java.base@11.0.4/StreamEncoder.java:104)
        - locked <0x00000000faf1aa50> (a java.io.OutputStreamWriter)
        at java.io.OutputStreamWriter.flushBuffer(java.base@11.0.4/OutputStreamWriter.java:184)
        at java.io.PrintStream.newLine(java.base@11.0.4/PrintStream.java:625)
        - locked <0x00000000faf18930> (a java.io.PrintStream)
        at java.io.PrintStream.println(java.base@11.0.4/PrintStream.java:883)
        - locked <0x00000000faf18930> (a java.io.PrintStream)
        at org.gradle.internal.io.LinePerThreadBufferingOutputStream.println(LinePerThreadBufferingOutputStream.java:203)
        at io.opentelemetry.javaagent.slf4j.impl.SimpleLogger.write(SimpleLogger.java:318)
        at io.opentelemetry.javaagent.slf4j.impl.SimpleLogger.log(SimpleLogger.java:295)
        at io.opentelemetry.javaagent.slf4j.impl.SimpleLogger.formatAndLog(SimpleLogger.java:355)
        at io.opentelemetry.javaagent.slf4j.impl.SimpleLogger.debug(SimpleLogger.java:446)
        at io.opentelemetry.javaagent.tooling.context.FieldBackedProvider$1$1$1.visitMethodInsn(FieldBackedProvider.java:201)
        at net.bytebuddy.jar.asm.MethodVisitor.visitMethodInsn(MethodVisitor.java:438)
        at net.bytebuddy.utility.visitor.ExceptionTableSensitiveMethodVisitor.onVisitMethodInsn(ExceptionTableSensitiveMethodVisitor.java:199)
        at net.bytebuddy.utility.visitor.ExceptionTableSensitiveMethodVisitor.visitMethodInsn(ExceptionTableSensitiveMethodVisitor.java:186)
        at net.bytebuddy.jar.asm.MethodVisitor.visitMethodInsn(MethodVisitor.java:438)
        at net.bytebuddy.utility.visitor.FramePaddingMethodVisitor.visitMethodInsn(FramePaddingMethodVisitor.java:94)
        at net.bytebuddy.jar.asm.MethodVisitor.visitMethodInsn(MethodVisitor.java:438)
        at net.bytebuddy.utility.visitor.StackAwareMethodVisitor.visitMethodInsn(StackAwareMethodVisitor.java:302)
        at net.bytebuddy.jar.asm.MethodVisitor.visitMethodInsn(MethodVisitor.java:438)
        at net.bytebuddy.utility.visitor.StackAwareMethodVisitor.visitMethodInsn(StackAwareMethodVisitor.java:302)
        at net.bytebuddy.jar.asm.MethodVisitor.visitMethodInsn(MethodVisitor.java:438)
        at net.bytebuddy.jar.asm.MethodVisitor.visitMethodInsn(MethodVisitor.java:438)
        at net.bytebuddy.jar.asm.ClassReader.readCode(ClassReader.java:2412)
        at net.bytebuddy.jar.asm.ClassReader.readMethod(ClassReader.java:1481)
        at net.bytebuddy.jar.asm.ClassReader.accept(ClassReader.java:711)
        at net.bytebuddy.jar.asm.ClassReader.accept(ClassReader.java:394)
        at net.bytebuddy.asm.Advice$Dispatcher$Inlining$Resolved$AdviceMethodInliner.apply(Advice.java:7821)
        at net.bytebuddy.asm.Advice$AdviceVisitor.onAfterExceptionTable(Advice.java:9955)
        at net.bytebuddy.utility.visitor.ExceptionTableSensitiveMethodVisitor.considerEndOfExceptionTable(ExceptionTableSensitiveMethodVisitor.java:49)
        at net.bytebuddy.utility.visitor.ExceptionTableSensitiveMethodVisitor.visitLabel(ExceptionTableSensitiveMethodVisitor.java:81)
        at net.bytebuddy.jar.asm.Label.accept(Label.java:358)
        at net.bytebuddy.jar.asm.ClassReader.readCode(ClassReader.java:2021)
        at net.bytebuddy.jar.asm.ClassReader.readMethod(ClassReader.java:1481)
        at net.bytebuddy.jar.asm.ClassReader.accept(ClassReader.java:711)
        at net.bytebuddy.jar.asm.ClassReader.accept(ClassReader.java:394)
        at net.bytebuddy.dynamic.scaffold.TypeWriter$Default$ForInlining.create(TypeWriter.java:3892)
        at net.bytebuddy.dynamic.scaffold.TypeWriter$Default.make(TypeWriter.java:2166)
        at net.bytebuddy.dynamic.scaffold.inline.RedefinitionDynamicTypeBuilder.make(RedefinitionDynamicTypeBuilder.java:224)
        at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.doTransform(AgentBuilder.java:10366)
        at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.transform(AgentBuilder.java:10302)
        at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.access$1600(AgentBuilder.java:10068)
        at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$Java9CapableVmDispatcher.run(AgentBuilder.java:10761)
        at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$Java9CapableVmDispatcher.run(AgentBuilder.java:10699)
        at java.security.AccessController.doPrivileged(java.base@11.0.4/Native Method)
        at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.transform(AgentBuilder.java:10258)
        at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$ByteBuddy$ModuleSupport.transform(Unknown Source)
        at sun.instrument.TransformerManager.transform(java.instrument@11.0.4/TransformerManager.java:188)
        at sun.instrument.InstrumentationImpl.transform(java.instrument@11.0.4/InstrumentationImpl.java:563)
        at java.lang.ClassLoader.defineClass1(java.base@11.0.4/Native Method)
        at java.lang.ClassLoader.defineClass(java.base@11.0.4/ClassLoader.java:1016)
        at java.security.SecureClassLoader.defineClass(java.base@11.0.4/SecureClassLoader.java:174)
        at java.net.URLClassLoader.defineClass(java.base@11.0.4/URLClassLoader.java:550)
        at java.net.URLClassLoader$1.run(java.base@11.0.4/URLClassLoader.java:458)
        at java.net.URLClassLoader$1.run(java.base@11.0.4/URLClassLoader.java:452)
        at java.security.AccessController.doPrivileged(java.base@11.0.4/Native Method)
        at java.net.URLClassLoader.findClass(java.base@11.0.4/URLClassLoader.java:451)
        at java.lang.ClassLoader.loadClass(java.base@11.0.4/ClassLoader.java:588)
        - locked <0x00000000fae3f740> (a java.lang.Object)
        at java.lang.ClassLoader.loadClass(java.base@11.0.4/ClassLoader.java:521)
        at org.gradle.internal.concurrent.ThreadFactoryImpl.newThread(ThreadFactoryImpl.java:36)
        at java.util.concurrent.ThreadPoolExecutor$Worker.<init>(java.base@11.0.4/ThreadPoolExecutor.java:623)
        at java.util.concurrent.ThreadPoolExecutor.addWorker(java.base@11.0.4/ThreadPoolExecutor.java:912)
        at java.util.concurrent.ThreadPoolExecutor.execute(java.base@11.0.4/ThreadPoolExecutor.java:1354)
        at org.gradle.internal.concurrent.ManagedExecutorImpl.execute(ManagedExecutorImpl.java:39)
        at org.gradle.internal.remote.internal.hub.MessageHub.addHandler(MessageHub.java:133)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection.addIncoming(MessageHubBackedObjectConnection.java:100)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker.startReceivingTests(TestWorker.java:105)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:69)
        at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:47)
        at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
        at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
        at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
        at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
        at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
"IntervalMetricReader-1":
        at jdk.internal.misc.Unsafe.park(java.base@11.0.4/Native Method)
        - parking to wait for  <0x00000000fba0b0b0> (a java.util.concurrent.locks.ReentrantLock$NonfairSync)
        at java.util.concurrent.locks.LockSupport.park(java.base@11.0.4/LockSupport.java:194)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.parkAndCheckInterrupt(java.base@11.0.4/AbstractQueuedSynchronizer.java:885)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquireQueued(java.base@11.0.4/AbstractQueuedSynchronizer.java:917)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@11.0.4/AbstractQueuedSynchronizer.java:1240)
        at java.util.concurrent.locks.ReentrantLock.lock(java.base@11.0.4/ReentrantLock.java:267)
        at org.gradle.internal.remote.internal.hub.MessageHub$ChannelDispatch.dispatch(MessageHub.java:370)
        at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
        at com.sun.proxy.$Proxy10.sendOutputEvent(Unknown Source)
        at org.gradle.process.internal.worker.child.WorkerLogEventListener.onOutput(WorkerLogEventListener.java:58)
        at jdk.internal.reflect.GeneratedMethodAccessor9.invoke(Unknown Source)
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(java.base@11.0.4/DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(java.base@11.0.4/Method.java:566)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:43)
        at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:245)
        at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:157)
        at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:141)
        at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
        at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
        at com.sun.proxy.$Proxy9.onOutput(Unknown Source)
        at org.gradle.internal.logging.sink.OutputEventTransformer.onOutput(OutputEventTransformer.java:104)
        at org.gradle.internal.logging.sink.OutputEventRenderer.onOutput(OutputEventRenderer.java:431)
        - locked <0x00000000fd873d20> (a java.lang.Object)
        at org.gradle.internal.logging.sink.OutputEventListenerManager$1.onOutput(OutputEventListenerManager.java:36)
        at org.gradle.internal.logging.services.TextStreamOutputEventListener.onTextEvent(TextStreamOutputEventListener.java:58)
        at org.gradle.internal.logging.services.TextStreamOutputEventListener.onOutput(TextStreamOutputEventListener.java:42)
        at org.gradle.internal.logging.source.PrintStreamLoggingSystem$OutputEventDestination.onOutput(PrintStreamLoggingSystem.java:173)
        at org.gradle.internal.logging.source.PrintStreamLoggingSystem$1.text(PrintStreamLoggingSystem.java:45)
        at org.gradle.internal.io.LineBufferingOutputStream.flush(LineBufferingOutputStream.java:98)
        at org.gradle.internal.io.LineBufferingOutputStream.write(LineBufferingOutputStream.java:82)
        at java.io.OutputStream.write(java.base@11.0.4/OutputStream.java:157)
        at java.io.PrintStream.write(java.base@11.0.4/PrintStream.java:559)
        - locked <0x00000000faff6630> (a java.io.PrintStream)
        at sun.nio.cs.StreamEncoder.writeBytes(java.base@11.0.4/StreamEncoder.java:233)
        at sun.nio.cs.StreamEncoder.implFlushBuffer(java.base@11.0.4/StreamEncoder.java:312)
        at sun.nio.cs.StreamEncoder.flushBuffer(java.base@11.0.4/StreamEncoder.java:104)
        - locked <0x00000000faff8750> (a java.io.OutputStreamWriter)
        at java.io.OutputStreamWriter.flushBuffer(java.base@11.0.4/OutputStreamWriter.java:184)
        at java.io.PrintStream.newLine(java.base@11.0.4/PrintStream.java:625)
        - locked <0x00000000faff6630> (a java.io.PrintStream)
        at java.io.PrintStream.println(java.base@11.0.4/PrintStream.java:883)
        - locked <0x00000000faff6630> (a java.io.PrintStream)
        at org.gradle.internal.io.LinePerThreadBufferingOutputStream.println(LinePerThreadBufferingOutputStream.java:203)
        at io.opentelemetry.javaagent.slf4j.impl.SimpleLogger.write(SimpleLogger.java:318)
        at io.opentelemetry.javaagent.slf4j.impl.SimpleLogger.log(SimpleLogger.java:295)
        at io.opentelemetry.javaagent.slf4j.impl.SimpleLogger.formatAndLog(SimpleLogger.java:355)
        at io.opentelemetry.javaagent.slf4j.impl.SimpleLogger.debug(SimpleLogger.java:446)
        at io.opentelemetry.javaagent.instrumentation.javaconcurrent.FutureInstrumentation$1.matches(FutureInstrumentation.java:85)
        at io.opentelemetry.javaagent.instrumentation.javaconcurrent.FutureInstrumentation$1.matches(FutureInstrumentation.java:80)
        at net.bytebuddy.matcher.ElementMatcher$Junction$Conjunction.matches(ElementMatcher.java:122)
        at io.opentelemetry.javaagent.tooling.bytebuddy.matcher.LoggingFailSafeMatcher.matches(LoggingFailSafeMatcher.java:52)
        at net.bytebuddy.agent.builder.AgentBuilder$RawMatcher$ForElementMatchers.matches(AgentBuilder.java:1312)
        at net.bytebuddy.agent.builder.AgentBuilder$RawMatcher$Conjunction.matches(AgentBuilder.java:1163)
        at net.bytebuddy.agent.builder.AgentBuilder$RawMatcher$Conjunction.matches(AgentBuilder.java:1163)
        at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.doTransform(AgentBuilder.java:10343)
        at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.transform(AgentBuilder.java:10302)
        at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.access$1600(AgentBuilder.java:10068)
        at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$Java9CapableVmDispatcher.run(AgentBuilder.java:10761)
        at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$Java9CapableVmDispatcher.run(AgentBuilder.java:10699)
        at java.security.AccessController.doPrivileged(java.base@11.0.4/Native Method)
        at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer.transform(AgentBuilder.java:10258)
        at net.bytebuddy.agent.builder.AgentBuilder$Default$ExecutingTransformer$ByteBuddy$ModuleSupport.transform(Unknown Source)
        at sun.instrument.TransformerManager.transform(java.instrument@11.0.4/TransformerManager.java:188)
        at sun.instrument.InstrumentationImpl.transform(java.instrument@11.0.4/InstrumentationImpl.java:563)
        at java.util.concurrent.CompletableFuture.uniHandleStage(java.base@11.0.4/CompletableFuture.java:944)
        at java.util.concurrent.CompletableFuture.handle(java.base@11.0.4/CompletableFuture.java:2266)
        at java.lang.ProcessImpl.initStreams(java.base@11.0.4/ProcessImpl.java:389)
        at java.lang.ProcessImpl.lambda$new$0(java.base@11.0.4/ProcessImpl.java:352)
        at java.lang.ProcessImpl$$Lambda$116/0x00000001002f9c40.run(java.base@11.0.4/Unknown Source)
        at java.security.AccessController.doPrivileged(java.base@11.0.4/Native Method)
        at java.lang.ProcessImpl.<init>(java.base@11.0.4/ProcessImpl.java:351)
        at java.lang.ProcessImpl.start(java.base@11.0.4/ProcessImpl.java:271)
        at java.lang.ProcessBuilder.start(java.base@11.0.4/ProcessBuilder.java:1107)
        at java.lang.ProcessBuilder.start(java.base@11.0.4/ProcessBuilder.java:1071)
        at java.lang.Runtime.exec(java.base@11.0.4/Runtime.java:591)
        at java.lang.Runtime.exec(java.base@11.0.4/Runtime.java:415)
        at java.lang.Runtime.exec(java.base@11.0.4/Runtime.java:312)
        at com.sun.jna.NativeLibrary.getLinuxLdPaths(NativeLibrary.java:999)
        at com.sun.jna.NativeLibrary.<clinit>(NativeLibrary.java:941)
        at com.sun.jna.Library$Handler.<init>(Library.java:192)
        at com.sun.jna.Native.load(Native.java:596)
        at com.sun.jna.Native.load(Native.java:570)
        at com.sun.jna.platform.linux.Udev.<clinit>(Udev.java:37)
        at oshi.hardware.platform.linux.LinuxHWDiskStore.getDisks(LinuxHWDiskStore.java:160)
        at oshi.hardware.platform.linux.LinuxHWDiskStore.getDisks(LinuxHWDiskStore.java:151)
        at oshi.hardware.platform.linux.LinuxHardwareAbstractionLayer.getDiskStores(LinuxHardwareAbstractionLayer.java:75)
        at io.opentelemetry.javaagent.shaded.instrumentation.oshi.SystemMetrics.lambda$registerObservers$6(SystemMetrics.java:138)
        at io.opentelemetry.javaagent.shaded.instrumentation.oshi.SystemMetrics$$Lambda$108/0x00000001002c4840.accept(Unknown Source)
        at io.opentelemetry.sdk.metrics.AsynchronousInstrumentAccumulator.lambda$longAsynchronousAccumulator$3(AsynchronousInstrumentAccumulator.java:59)
        at io.opentelemetry.sdk.metrics.AsynchronousInstrumentAccumulator$$Lambda$98/0x00000001002c2040.accept(Unknown Source)
        at io.opentelemetry.sdk.metrics.AsynchronousInstrumentAccumulator.collectAll(AsynchronousInstrumentAccumulator.java:31)
        at io.opentelemetry.sdk.metrics.AbstractAsynchronousInstrument.collectAll(AbstractAsynchronousInstrument.java:23)
        at io.opentelemetry.sdk.metrics.SdkMeter.collectAll(SdkMeter.java:102)
        at io.opentelemetry.sdk.metrics.SdkMeterProvider$MetricProducerSdk.collectAllMetrics(SdkMeterProvider.java:174)
        at io.opentelemetry.sdk.metrics.export.IntervalMetricReader$Exporter.run(IntervalMetricReader.java:192)
        at java.util.concurrent.Executors$RunnableAdapter.call(java.base@11.0.4/Executors.java:515)
        at java.util.concurrent.FutureTask.runAndReset(java.base@11.0.4/FutureTask.java:305)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(java.base@11.0.4/ScheduledThreadPoolExecutor.java:305)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@11.0.4/ThreadPoolExecutor.java:1128)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@11.0.4/ThreadPoolExecutor.java:628)
        at java.lang.Thread.run(java.base@11.0.4/Thread.java:834)

Found 1 deadlock.
```

Reply:
Have you reported that to Gradle?

Reply:
no, I can report it tomorrow

Reply:
been investigating, not a gradle bug, the javaagent is conspiring to cause the deadlock, finally found a way to resolve it (along with lots of ways not to resolve it 😭), I still need to write up explanation and add comments, will do that later today

Reply:
spoke too soon 😭😭😭

still hanging on CI sporadically, need to investigate more

The basic problem is:

1. gradle does some locking, assuming it knows what happens under those locks

2. gradle accesses some classes that haven't been loaded yet, triggering class loading underneath those locks

3. during that class loading, we perform some debug logging (particularly in `FieldBackedProvider`)

4. but we are logging to `System.out`, and gradle has hijacked `System.out`, so that logging goes back to gradle

5. where gradle does more locking

6. [add multiple threads]

Looking at heap dumps from the deadlocked state shows what class gradle was accessing that was triggering class loading, which is what led me to add `org.gradle.` to the `GlobalIgnoresMatcher` and to preload `io.opentelemetry.javaagent.instrumentation.api.concurrent.RunnableWrapper` during agent startup, so gradle wouldn't be causing it to be loaded (see https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1643/commits/e39246e78cc436ec7ed98465d90985a46d902bb3).

Reply:
I reverted the deadlock related changes from today so that it's green for @iNikem to review

I'll continue looking into the deadlock issue tomorrow (I can either continue in this PR, or leave debug logging disabled in this PR and work on re-enabling it in a separate PR)

Reply:
I think https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1643/commits/839c55e2e4a9c2907f42894167dde2c9c6c91aa1 is a reasonable solution (now part of this PR).

These were the three classes that I saw being instrumented when deadlocking occurred:
* `io.opentelemetry.javaagent.instrumentation.api.concurrent.RunnableWrapper`
* `java.util.concurrent.ForkJoinPool`
* `org.gradle.something` (I don't remember - but this PR adds `org.gradle` to the GlobalIgnoresMatcher so this one is not a problem anymore)